### PR TITLE
Add support for multiple inputs

### PR DESCRIPTION
**Pull request by rshym**

### CODE REVIEW

File: device.c [hunk 1]
```diff
@@ -39,10 +39,30 @@
...
+#define NUMBER_OF_DEVICES_IN 3
+
 /* Number of supported output devices. Should be in sync with
  * number of output buses in audio_policy_configuration.xml */
 #define NUMBER_OF_DEVICES_OUT 8
 
+
+typedef struct xa_device_map
+{
+    audio_devices_t device_type;  /* something like AUDIO_DEVICE_OUT_BUS */
```
Comment (andr2000) [inline device.c]: I would put 'one of the AUDIO_DEVICE_OUT_XXX'

File: device.c [hunk 2]
```diff
@@ -39,10 +39,30 @@
...
+xa_device_map_t xa_input_map[NUMBER_OF_DEVICES_IN] =
+{
+    {AUDIO_DEVICE_IN_BUILTIN_MIC, 0, 0},
+    {AUDIO_DEVICE_IN_LINE, 0, 0},
+    {AUDIO_DEVICE_IN_BUS, 0, 0},
+};
+
+
 typedef struct x_audio_device {
     /* NOTE audio_hw_device_t MUST be very first member of structure */
```
Comment (andr2000) [inline device.c]: NOTE**:**

File: device.c [hunk 3]
```diff
@@ -226,17 +249,20 @@ int adev_set_mode(struct audio_hw_device *dev, audio_mode_t mode)
 int adev_set_mic_mute(struct audio_hw_device *dev, bool state)
 {
     x_audio_device_t *xdev = (x_audio_device_t*)dev;
+    unsigned int i = 0;
```
Comment (andr2000) [inline device.c]: no need to assign 0 here

File: device.c [hunk 4]
```diff
@@ -428,25 +453,52 @@ int adev_open_input_stream(struct audio_hw_device *dev,
...
     if ((devices & AUDIO_DEVICE_BIT_IN) == 0) {
         ALOGE("Failed. Incorrect device type. -EINVAL");
         *stream_in = NULL;
         pthread_mutex_unlock(&xdev->lock);
         return -EINVAL;
     }
 
-    /* TODO define card_id and dev_id for stream, using provided parameters:
-       devices, flags, address, source */
+    unsigned int slot = 0;
```
Comment (andr2000) [inline device.c]: no need for 0 assignment

File: device.c [hunk 5]
```diff
@@ -428,25 +453,52 @@ int adev_open_input_stream(struct audio_hw_device *dev,
...
+    /* For now we have one card, so card id is always constant */
```
Comment (andr2000) [inline device.c]: add FIXME here

File: device.c (hunk 1 above)
Comment (andr2000) [inline device.c]: device_type_mask?

File: device.c [hunk 6]
```diff
@@ -428,25 +453,52 @@ int adev_open_input_stream(struct audio_hw_device *dev,
...
+    /* Get device id using device type */
+    for (slot = 0; slot < NUMBER_OF_DEVICES_IN; slot++) {
+        if ((devices & xa_input_map[slot].device_type) != 0) {
+            if (xa_input_map[slot].device_type == AUDIO_DEVICE_IN_BUS) {
+                if (sscanf(address, "bus%u", &dev_id) == 1) {
+                    /* dev_id is identified */
+                    dev_id = xa_input_map[slot].pcm_dev_id;
+                } else {
+                    ALOGE("Address has not supported format.");
```
Comment (andr2000) [inline device.c]: 'Address has unsupported format'?

File: device.c [hunk 7]
```diff
@@ -428,25 +453,52 @@ int adev_open_input_stream(struct audio_hw_device *dev,
...
+                    dev_id = NUMBER_OF_DEVICES_IN;
+                }
+            } else {
+                /* dev_id is identified */
+                dev_id = xa_input_map[slot].pcm_dev_id;
```
Comment (andr2000) [inline device.c]: you use slot to check if dev_id was found, so it is not needed to set dev_id to wrong value here

File: device.c [hunk 8]
```diff
@@ -428,25 +453,52 @@ int adev_open_input_stream(struct audio_hw_device *dev,
...
+            res = in_create(dev, handle, devices, CARD_ID, dev_id, config, stream_in);
+            if (*stream_in != NULL) {
+                xdev->xin_streams[slot] = (x_stream_in_t*)(*stream_in);
+                if (xdev->mic_is_muted) {
+                    in_set_mute(xdev->xin_streams[slot], true);
+                }
+            }
+            LOG_FN_PARAMETERS("Created stream_in:%p", *stream_in);
+        } else {
+            res = -ENOMEM;
```
Comment (andr2000) [inline device.c]: why ENOMEM? EEXISTS?

File: device.c [hunk 9]
```diff
@@ -455,17 +507,22 @@ int adev_open_input_stream(struct audio_hw_device *dev,
 void adev_close_input_stream(struct audio_hw_device *dev, struct audio_stream_in *stream_in)
 {
     x_audio_device_t *xdev = (x_audio_device_t*)dev;
+    unsigned int i = 0;
```
Comment (andr2000) [inline device.c]: no need to assign to 0 here

File: device.c [hunk 10]
```diff
@@ -482,7 +539,7 @@ int adev_get_microphones(const struct audio_hw_device *dev,
 int adev_dump(const struct audio_hw_device *dev, int fd)
 {
     x_audio_device_t *xdev = (x_audio_device_t*)dev;
-    int i = 0;
+    unsigned int i = 0;
```
Comment (andr2000) [inline device.c]: ditto

File: device.c (hunk 1 above)
Comment (rshym) [reply device.c]: Done.

File: device.c (hunk 3 above)
Comment (rshym) [reply device.c]: Done.

File: device.c (hunk 8 above)
Comment (rshym) [reply device.c]: Done.

File: device.c (hunk 9 above)
Comment (rshym) [reply device.c]: Done.

File: device.c (hunk 10 above)
Comment (rshym) [reply device.c]: Done.

File: device.c (hunk 2 above)
Comment (rshym) [reply device.c]: Done.

File: device.c (hunk 5 above)
Comment (rshym) [reply device.c]: Comment removed because card index was added to configuration table.

File: device.c (hunk 6 above)
Comment (rshym) [reply device.c]: Log message removed because this case is considered not as error but as 'non matching'.

File: device.c (hunk 4 above)
Comment (rshym) [reply device.c]: Done.

File: device.c (hunk 1 above)
Comment (rshym) [reply device.c]: Done.

File: device.c (hunk 7 above)
Comment (rshym) [reply device.c]: Fixed.

**Review by andr2000 — APPROVED** [review]
`Reviewed-by: Oleksandr Andrushchenko <oleksandr_andrushchenko@epam.com>`

**Pull request merged**